demo/test/cost_estimator.cpp

```diff
@@ -0,0 +1,74 @@
+#include "demo/cost_estimator.hpp"
+
+#include <gtest/gtest.h>
+
+#include "mock_environment_model.hpp"
+
+namespace demo {
+
+class CostEstimatorTest : public ::testing::Test {
+protected:
+    CostEstimatorTest()
+            : environmentModel_(std::make_shared<MockEnvironmentModel>()),
+              costEstimator_{environmentModel_, parameters_} {
+        setMaze();
+        environmentModel_->setGhostPositions({1, 1});
+    }
+
+    void setMaze() {
+        const char str[] = {"##########"
+                            "#        #"
+                            "#        #"
+                            "#        #"
+                            "#        #"
+                            "#.       #"
+                            "#.       #"
+                            "#...     #"
+                            "#...     #"
+                            "##########"};
+        environmentModel_->setMaze({10, 10}, str);
+    }
+
+    MockEnvironmentModel::Ptr environmentModel_;
+    CostEstimator::Parameters parameters_{1};
+
+    CostEstimator costEstimator_;
+};
+
+TEST_F(CostEstimatorTest, estimateCostDownPath) {
+    // Move from the top left to the bottom left corner
+    environmentModel_->setPacmanPosition({1, 1});
+    using D = Direction;
+    Path path{D::DOWN, D::DOWN, D::DOWN, D::DOWN, D::DOWN, D::DOWN, D::DOWN};
+
+    // Number of cells along the path and within the neighborhood search area
+    int numOfCells = 8 + 9;
+
+    // Number of dots along the path and within the neighborhood search area
+    int numOfDots = 4 + 4;
+
+    double expectedCost = static_cast<double>(numOfCells) / numOfDots;
+    double actualCost = costEstimator_.estimateCost(Command{path}, true);
+
+    EXPECT_EQ(expectedCost, actualCost);
+}
+
+TEST_F(CostEstimatorTest, estimateCostUpPath) {
+    // Move from the bottom left to the top left corner
+    environmentModel_->setPacmanPosition({1, 8});
+    using D = Direction;
+    Path path{D::UP, D::UP, D::UP, D::UP, D::UP, D::UP, D::UP};
+
+    // Number of cells along the path and within the neighborhood search area
+    int numOfCells = 8 + 9;
```

Comment:
Well the question is whether you count the current position as part of the path which is the case right now. But giving it some more thought I would probably change that. The relative path doesn't contain the current position (it can't) so consequently the absolute path probably shouldn't either.

Reply:
Fixed in e48d8ba73c3c